simulation/halsim_ws_core/src/main/native/cpp/WSProvider_Analog.cpp

```diff
@@ -0,0 +1,70 @@
+/*----------------------------------------------------------------------------*/
+/* Copyright (c) 2017-2020 FIRST. All Rights Reserved.                        */
+/* Open Source Software - may be modified and shared by FRC teams. The code   */
+/* must be accompanied by the FIRST BSD license file in the root directory of */
+/* the project.                                                               */
+/*----------------------------------------------------------------------------*/
+
+#include "WSProvider_Analog.h"
+
+#include <hal/Ports.h>
+#include <hal/simulation/AnalogInData.h>
+#include <hal/simulation/AnalogOutData.h>
+
+namespace wpilibws {
+
+void HALSimWSProviderAnalogIn::Initialize(WSRegisterFunc webRegisterFunc) {
+  CreateProviders<HALSimWSProviderAnalogIn>("AI", HAL_GetNumAnalogInputs(),
+                                            HALSIM_RegisterAnalogInAllCallbacks,
+                                            webRegisterFunc);
+}
+
+wpi::json HALSimWSProviderAnalogIn::OnSimValueChanged(const char* cbName) {
+  return {
+      {"<init", static_cast<bool>(HALSIM_GetAnalogInInitialized(m_channel))},
+      {"<avg_bits", HALSIM_GetAnalogInAverageBits(m_channel)},
+      {"<oversample_bits", HALSIM_GetAnalogInOversampleBits(m_channel)},
+      {">voltage", HALSIM_GetAnalogInVoltage(m_channel)},
+      {
+          "accum",
+          {
+              {"<init",
+               static_cast<bool>(
+                   HALSIM_GetAnalogInAccumulatorInitialized(m_channel))},
+              {">value", HALSIM_GetAnalogInAccumulatorValue(m_channel)},
+              {">count", HALSIM_GetAnalogInAccumulatorCount(m_channel)},
+              {"<center", HALSIM_GetAnalogInAccumulatorCenter(m_channel)},
+              {"<deadband", HALSIM_GetAnalogInAccumulatorDeadband(m_channel)},
+          },
+      },
+  };
+}
+
+void HALSimWSProviderAnalogIn::OnNetValueChanged(const wpi::json& json) {
+  wpi::json::const_iterator it;
+  if ((it = json.find(">voltage")) != json.end()) {
+    HALSIM_SetAnalogInVoltage(m_channel, it.value());
+  }
+  if ((it = json.find("accum")) != json.end()) {
+    auto accum = it.value();
+    if ((it = accum.find(">value")) != accum.end()) {
+      HALSIM_SetAnalogInAccumulatorValue(m_channel, it.value());
+    }
+    if ((it = json.find(">count")) != json.end()) {
+      HALSIM_SetAnalogInAccumulatorCount(m_channel, it.value());
+    }
+  }
+}
+
+void HALSimWSProviderAnalogOut::Initialize(WSRegisterFunc webRegisterFunc) {
+  CreateProviders<HALSimWSProviderAnalogOut>(
+      "AO", HAL_GetNumAnalogOutputs(), HALSIM_RegisterAnalogOutAllCallbacks,
+      webRegisterFunc);
```

Comment:
For efficiency reasons we should use the specific callbacks rather than using register all callbacks and doing a diff on every update.  The callbacks are called synchronously with each change so the WS code doing the json work is actually going to be blocking user code. (e.g. User calls HAL which calls the WS code).

Reply:
@PeterJohnson since the `OnSimValueChanged` function already takes in a `char*` representing the callback type, would it be alright to continue using the RegisterAllCallbacks functionality, but return the appropriate subset of data there (this way we don't have to run the diff)

Reply:
I updated the providers to only send appropriate data based on the callback type (e.g. the `Initialized` callback type will only send `<init` fields)